src/connector/src/sink/iceberg/mod.rs

```diff
@@ -1963,18 +2080,32 @@ impl SinglePhaseCommitCoordinator for IcebergSinkCommitter {
         add_columns: Option<Vec<Field>>,
     ) -> Result<()> {
         tracing::info!("Starting iceberg direct commit in epoch {epoch}");
-
+        // Commit data if present
         let (write_results, snapshot_id) =
-            match self.pre_commit_inner(epoch, metadata, add_columns)? {
+            match self.pre_commit_inner(epoch, metadata, add_columns.clone())? {
                 Some((write_results, snapshot_id)) => (write_results, snapshot_id),
                 None => {
-                    tracing::debug!(?epoch, "no data to commit");
+                    // No data to commit, but may have schema change
+                    if let Some(add_columns) = add_columns {
```

Comment:
Also apply those change for non exactly once situation, and it's much simpler than exactly once.